File: app/src/main/java/io/github/gmazzo/android/livewallpaper/weather/engine/things/ThingSun.java
package io.github.gmazzo.android.livewallpaper.weather.engine.things;

import static javax.microedition.khronos.opengles.GL10.GL_MODELVIEW;
import static javax.microedition.khronos.opengles.GL10.GL_MODULATE;
import static javax.microedition.khronos.opengles.GL10.GL_TEXTURE0;

import javax.microedition.khronos.opengles.GL10;
import javax.microedition.khronos.opengles.GL11;

import io.github.gmazzo.android.livewallpaper.weather.R;
import io.github.gmazzo.android.livewallpaper.weather.engine.EngineColor;
import io.github.gmazzo.android.livewallpaper.weather.engine.scenes.SceneBase;
import io.github.gmazzo.android.livewallpaper.weather.wallpaper.Models;
import io.github.gmazzo.android.livewallpaper.weather.wallpaper.Texture;
import io.github.gmazzo.android.livewallpaper.weather.wallpaper.Textures;

public class ThingSun extends Thing {
    private static final String TAG = "Sun";
    private Texture sunBlend;

    public ThingSun() {
        this.engineColor = new EngineColor(1.0f, 1.0f, 0.95f, 1.0f);
    }

    @Override
    public void render(GL10 gl, Textures textures, Models models) {
        if (texture == null) {
            texture = textures.get(R.raw.sun);
            sunBlend = textures.get(R.raw.sun_blend);
            model = models.get(R.raw.plane_16x16);
        }

        gl.glBlendFunc(1, 769);
        gl.glColor4f(this.engineColor.getR(), this.engineColor.getG(), this.engineColor.getB(), this.engineColor.getA());
        gl.glMatrixMode(GL_MODELVIEW);
        gl.glPushMatrix();
        gl.glLoadIdentity();
        gl.glTranslatef(this.origin.getX(), this.origin.getY(), this.origin.getZ());
        gl.glScalef(this.scale.getX(), this.scale.getY(), this.scale.getZ());
        gl.glRotatef((this.sTimeElapsed * 12.0f) % 360.0f, 0.0f, 1.0f, 0.0f);
        gl.glMatrixMode(5890);
        gl.glPushMatrix();
        float f11 = (this.sTimeElapsed * 18.0f) % 360.0f;
        gl.glTranslatef(0.5f, 0.5f, 0.0f);
        gl.glRotatef(f11, 0.0f, 0.0f, 1.0f);
        gl.glTranslatef(-0.5f, -0.5f, 0.0f);
        if (gl instanceof GL11) {
            model.renderFrameMultiTexture(sunBlend, texture, GL_MODULATE, false);
        } else {
            gl.glBindTexture(GL_TEXTURE0, texture.getGlId());
            model.render();
        }
        gl.glPopMatrix();
        gl.glMatrixMode(GL_MODELVIEW);
        gl.glPopMatrix();
    }

    public void update(float timeDelta) {
        super.update(timeDelta);
        float sunPos = SceneBase.todSunPosition;
        float alpha = 0.0f;
        if (sunPos > 0.0f) {
            this.scale.set(2.0f);
            float altitude = 175.0f * sunPos;
            alpha = altitude / 25.0f;
            if (alpha > 1.0f) {
                alpha = 1.0f;
            }
            this.origin.setZ(altitude - 50.0f);
            if (this.origin.getZ() > 40.0f) {
                this.origin.setZ(40.0f);
            }
        } else {
            this.scale.set(0.0f);
        }
        this.engineColor.set(SceneBase.todEngineColorFinal);
        this.engineColor.times(alpha);
    }
}

File: app/src/main/kotlin/io/github/gmazzo/android/livewallpaper/weather/engine/things/ThingSun.kt
package io.github.gmazzo.android.livewallpaper.weather.engine.things

import io.github.gmazzo.android.livewallpaper.weather.R
import io.github.gmazzo.android.livewallpaper.weather.engine.GlobalTime
import io.github.gmazzo.android.livewallpaper.weather.engine.Vector
import io.github.gmazzo.android.livewallpaper.weather.engine.models.Models
import io.github.gmazzo.android.livewallpaper.weather.engine.pushMatrix
import io.github.gmazzo.android.livewallpaper.weather.engine.textures.Textures
import io.github.gmazzo.android.livewallpaper.weather.engine.timeofday.TimeOfDay
import io.github.gmazzo.android.livewallpaper.weather.engine.withColor
import javax.inject.Inject
import javax.microedition.khronos.opengles.GL10.GL_MODELVIEW
import javax.microedition.khronos.opengles.GL10.GL_MODULATE
import javax.microedition.khronos.opengles.GL10.GL_ONE
import javax.microedition.khronos.opengles.GL10.GL_ONE_MINUS_SRC_COLOR
import javax.microedition.khronos.opengles.GL10.GL_TEXTURE
import javax.microedition.khronos.opengles.GL11
import kotlin.math.min

class ThingSun @Inject constructor(
    gl: GL11,
    models: Models,
    textures: Textures,
    private val time: GlobalTime,
    private val timeOfDay: TimeOfDay,
) : Thing(gl, models[R.raw.plane_16x16], textures[R.raw.sun]) {

    private val sunBlend = textures[R.raw.sun_blend]

    init {
        scale = Vector(2f)
        color.set(1f, 1f, .95f, 1f)
    }

    override fun render() = gl.pushMatrix(GL_MODELVIEW) {
        val timeElapsed = time.elapsedSeconds

        gl.glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_COLOR)
        gl.glTranslatef(origin.x, origin.y, origin.z)
        gl.glScalef(scale.x, scale.y, scale.z)
        gl.glRotatef((timeElapsed * 12f) % 360f, 0f, 1f, 0f)

        pushMatrix(GL_TEXTURE) {
            val angle = (timeElapsed * 18f) % 360f

            gl.glTranslatef(.5f, .5f, 0f)
            gl.glRotatef(angle, 0f, 0f, 1f)
            gl.glTranslatef(-.5f, -.5f, 0f)

            withColor(color) {
                model.renderFrameMultiTexture(sunBlend, texture, GL_MODULATE, false)
            }
        }
    }

    override fun update() {
        val altitude = 215f * timeOfDay.sunPosition

        origin = origin.copy(z = min(altitude - 90f, 0f))
        super.update()
    }

}
